### ISSUE TITLE
CA1508 false positive on "Create" methods in nullable context

### ISSUE THREAD

**Issue by CodeBlanch** [dotnet/roslyn-analyzers#4626]
### Analyzer

**Diagnostic ID**: [CA1508](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1508)

### Analyzer source

```xml
<PackageReference Include="Microsoft.CodeAnalysis.NetAnalyzers" Version="5.0.1">
```
### Describe the bug

There seems to be some cheating going on with CA1508 where if a method is named `Create*` the rule assumes the return will be non-null.

![image](https://user-images.githubusercontent.com/28367120/102831403-95ae7980-43a0-11eb-8bb0-47f7747b66f7.png)

That cheat I don't think makes a lot of sense in a nullable context where the nullability is explicit.

### Expected behavior

The code is indicating the return could be null so in this case, I think CA1508 should use that information instead of the `Create*` cheat and the warning should NOT fire.

## Additional context

Example of the issue:

```csharp
#nullable enable

		public class MyClass
		{
			public static MyClass? Create() => null;

			public static MyClass? Foo() => null;

			public static void Logic()
			{
				MyClass? instance = Create();
				if (instance == null) // CA1508 warning generated here
					throw new InvalidOperationException();

				MyClass? instance2 = Foo();
				if (instance2 == null) // No warning generated
					throw new InvalidOperationException();
			}
		}

#nullable disable
```

**Comment by Youssef1313** [comment]
@mavasani This was probably fixed in #4554
A test will be needed to confirm and prevent regressions.

**Comment by mavasani** [comment]
Agreed. @genlu can you verify with a unit test?

**Comment by genlu** [comment]
Verified with code above this is fixed. Given we already have a test go though exact same code path, I didn't add a new test for it.
https://github.com/dotnet/roslyn-analyzers/blob/1ead3c96987ec6200209ab602529b9bab573833d/src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/Maintainability/AvoidDeadConditionalCode_NullAnalysis.cs#L7013
